package com.predator.common.gameplay.item;

import com.predator.common.gameplay.entity.projectile.ShurikenProjectile;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.stats.Stats;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;

public class ShurikenItem extends Item {

    public ShurikenItem() {
        super(new Properties().stacksTo(1));
    }

    @Override
    public int getUseDuration(@NotNull ItemStack stack, @NotNull LivingEntity entity) {
        return 72000;
    }

    @Override
    public void releaseUsing(@NotNull ItemStack stack, @NotNull Level level, @NotNull LivingEntity livingEntity, int timeCharged) {
        int i = this.getUseDuration(stack, livingEntity) - timeCharged;
        float powerForTime = getPowerForTime(i);
        if (powerForTime > 0.1 && livingEntity instanceof Player player && !player.getCooldowns().isOnCooldown(this)) {
            player.getCooldowns().addCooldown(this, 5);
            // TODO: Change sound effect here.
            level.playSound(
                null,
                player.getX(),
                player.getY(),
                player.getZ(),
                SoundEvents.TRIDENT_THROW,
                SoundSource.PLAYERS,
                0.5F,
                0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
            );

            if (!level.isClientSide) {
                var shurikenItemEntity = new ShurikenProjectile(level, player);
                shurikenItemEntity.setItem(stack);
                shurikenItemEntity.setOwner(player);
                shurikenItemEntity.shootFromRotation(player, player.getXRot(), player.getYRot(), 0.0F, powerForTime * 6.5F, 1.0F);
                level.addFreshEntity(shurikenItemEntity);
            }
            player.awardStat(Stats.ITEM_USED.get(this));
            if (!player.getAbilities().instabuild)
                stack.shrink(1);
        }

        super.releaseUsing(stack, level, livingEntity, timeCharged);
    }

    public static float getPowerForTime(int charge) {
        var normalizedCharge = charge / 20.0F;
        normalizedCharge = (normalizedCharge * normalizedCharge + normalizedCharge * 2.0F) / 3.0F;
        if (normalizedCharge > 1.0F)
            normalizedCharge = 1.0F;
        return normalizedCharge;
    }

    @Override
    public @NotNull InteractionResultHolder<ItemStack> use(@NotNull Level level, Player player, @NotNull InteractionHand hand) {
        var itemStack = player.getItemInHand(hand);
        player.startUsingItem(hand);
        return InteractionResultHolder.consume(itemStack);
    }
}
